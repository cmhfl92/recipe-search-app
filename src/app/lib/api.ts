import axios from 'axios';
import { RecipeSearchResponse } from '@/app/types/recipe';

const APP_ID = process.env.NEXT_PUBLIC_EDAMAM_APP_ID!;
const APP_KEY = process.env.NEXT_PUBLIC_EDAMAM_APP_KEY!;
const BASE_URL = 'https://api.edamam.com/api/recipes/v2';

export const searchRecipes = async (
  query: string
): Promise<RecipeSearchResponse> => {
  const url = `${BASE_URL}?type=public&q=${encodeURIComponent(
    query
  )}&app_id=${APP_ID}&app_key=${APP_KEY}`;
  const response = await axios.get<RecipeSearchResponse>(url);
  return response.data;
};
