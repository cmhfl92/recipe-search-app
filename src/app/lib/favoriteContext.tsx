'use client';

import { createContext, useContext, useEffect, useState } from 'react';
import { useFormState } from 'react-dom';

type FavoritesContextType = {
  favorites: string[];
  addFavorite: (id: string) => void;
  removeFavorite: (id: string) => void;
  isFavorite: (id: string) => boolean;
};

const FavoritesContext = createContext<FavoritesContextType | undefined>(
  undefined
);

export const FavoritesProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const [favorites, setFavorites] = useState<string[]>([]);

  useEffect(() => {
    const stored = localStorage.getItem('favorites');
    if (stored) {
      setFavorites(JSON.parse(stored));
    }
  }, []);

  useEffect(() => {
    localStorage.setItem('favorites', JSON.stringify(favorites));
  }, [favorites]);

  const addFavorite = (id: string) => {
    setFavorites(prev => [...new Set([...prev, id])]);
  };

  const removeFavorite = (id: string) => {
    setFavorites(prev => prev.filter(fav => fav !== id));
  };

  const isFavorite = (id: string) => favorites.includes(id);

  return (
    <FavoritesContext.Provider
      value={{ favorites, addFavorite, removeFavorite, isFavorite }}
    >
      {children}
    </FavoritesContext.Provider>
  );
};

export const useFavorites = () => {
  const context = useContext(FavoritesContext);
  if (!context) {
    throw new Error('useFavorites must be used within a FavoritesProvider');
  }
  return context;
};
