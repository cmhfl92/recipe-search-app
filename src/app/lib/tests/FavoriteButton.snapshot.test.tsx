import React from 'react';
import { render } from '@testing-library/react';
import { FavoritesProvider, useFavorites } from '@/app/lib/favoriteContext';

const FavoriteButton = ({ recipeId }: { recipeId: string }) => {
  const { isFavorite, addFavorite, removeFavorite } = useFavorites();
  const favorite = isFavorite(recipeId);

  return (
    <button
      onClick={() =>
        favorite ? removeFavorite(recipeId) : addFavorite(recipeId)
      }
    >
      {favorite ? 'üíî' : '‚ù§Ô∏è'}
    </button>
  );
};

describe('FavoriteButton Snapshot', () => {
  const renderWithProvider = (ui: React.ReactNode) =>
    render(<FavoritesProvider>{ui}</FavoritesProvider>);

  it('renders non-favorited state', () => {
    const { asFragment } = renderWithProvider(
      <FavoriteButton recipeId='123' />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders favorited state after click', () => {
    const { asFragment, getByRole } = renderWithProvider(
      <FavoriteButton recipeId='123' />
    );
    getByRole('button').click();
    expect(asFragment()).toMatchSnapshot();
  });
});
