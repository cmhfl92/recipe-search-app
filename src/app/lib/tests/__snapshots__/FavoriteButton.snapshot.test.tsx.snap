// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteButton Snapshot renders favorited state after click 1`] = `
<DocumentFragment>
  <button>
    ❤️
  </button>
</DocumentFragment>
`;

exports[`FavoriteButton Snapshot renders non-favorited state 1`] = `
<DocumentFragment>
  <button>
    ❤️
  </button>
</DocumentFragment>
`;
